### PR TITLE
Add new ConfigDocument API methods

### PR DESCRIPTION
**Pull request by fpringvaldsen**
Add two new methods to the ConfigDocument API, hasValue() and
removeValue(), along with tests.

### CODE REVIEW

File: config/src/main/java/com/typesafe/config/ConfigDocument.java [hunk 1]
```diff
@@ -51,6 +51,24 @@
...
     /**
+     * Returns a new ConfigDocument that is a copy of the current ConfigDocument, but with
+     * the value at the desired path removed. If the desired path does not exist in the document,
+     * a copy of the current document will be returned. If there is an array at the root, an exception
+     * will be thrown.
+     *
+     * @param path the path to remove from the document
+     * @return a copy of the ConfigDocument with the desired value removed from the document.
+     */
+    ConfigDocument removeValue(String path);
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/ConfigDocument.java]: docs should say "all values at the desired path" probably (if there are multiple, we remove all presumably)

**Comment by fpringvaldsen** [comment]
I need to rebase and make a few other small changes, but this should at least be ready for review now.

File: config/src/main/java/com/typesafe/config/ConfigDocument.java [hunk 2]
```diff
@@ -51,6 +51,24 @@
...
+
+    /**
+     * Returns a boolean indicating whether or not a ConfigDocument has a value at the desired path.
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/ConfigDocument.java]: maybe add "Substitutions (the `<code>${foo}</code>` syntax) are always considered to be values for purposes of this method, even though they might disappear once the file is resolved." 
Or something like that.

File: config/src/main/java/com/typesafe/config/impl/ConfigNodeObject.java [hunk 3]
```diff
@@ -12,32 +12,83 @@
         super(children);
     }
 
-    protected ConfigNodeObject changeValueOnPath(Path desiredPath, AbstractConfigNodeValue value) {
+    public boolean hasValue(Path desiredPath) {
+        for (AbstractConfigNode node : children) {
+            if (node instanceof ConfigNodeField) {
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/impl/ConfigNodeObject.java]: if you put `ConfigNodeField field = (ConfigNodeField) node` here and then use `field` below, then you could avoid some of the cast noise in the code below.

File: config/src/main/java/com/typesafe/config/impl/ConfigNodeObject.java [hunk 4]
```diff
@@ -12,32 +12,83 @@
...
+                Path key = ((ConfigNodeField) node).path().value();
+                if (key.equals(desiredPath) || key.startsWith(desiredPath)) {
+                    return true;
+                } else if (desiredPath.startsWith(key)) {
+                    if (((ConfigNodeField) node).value() instanceof ConfigNodeObject) {
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/impl/ConfigNodeObject.java]: similarly here you could write `ConfigNodeObject obj = (ConfigNodeObject) field.value()` to avoid mixing cast noise into the logic below.

File: config/src/main/java/com/typesafe/config/impl/ConfigNodeObject.java [hunk 5]
```diff
@@ -12,32 +12,83 @@
...
-            if (!(super.children.get(i) instanceof ConfigNodeField)) {
+        for (int i = childrenCopy.size() - 1; i >= 0; i--) {
+            if (childrenCopy.get(i) instanceof ConfigNodeSingleToken) {
+                Token t = ((ConfigNodeSingleToken) childrenCopy.get(i)).token();
+                // Ensure that, when we are removing settings in JSON, we don't end up with a trailing comma
+                if (flavor == ConfigSyntax.JSON && !seenNonMatching && t == Tokens.COMMA) {
+                    childrenCopy.remove(i);
+                }
+                continue;
+            }else if (!(childrenCopy.get(i) instanceof ConfigNodeField)) {
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/impl/ConfigNodeObject.java]: missing a space after the `}`

File: config/src/test/scala/com/typesafe/config/impl/ConfigDocumentTest.scala [hunk 6]
```diff
@@ -193,6 +225,28 @@ class ConfigDocumentTest extends TestUtils {
...
+        exceptionThrown = false;
+        try {
+            document.hasValue("a")
+        } catch {
+            case e: Exception =>
+                exceptionThrown = true
+                assertTrue(e.isInstanceOf[ConfigException])
+                assertTrue(e.getMessage.contains("ConfigDocument had an array at the root level"))
+        }
+        assertTrue(exceptionThrown)
```
Comment (havocp) [inline config/src/test/scala/com/typesafe/config/impl/ConfigDocumentTest.scala]: you can use `val e = intercept[ConfigException] { document.hasValue("a") }; assertTrue(e.getMessage.contains)` here, which will check the same things with less code. (intercept[E] throws if the exception is not intercepted or doesn't have the expected type)

**Comment by havocp** [comment]
will need fixing up since I moved files around, too. (sorry about that)

**Comment by fpringvaldsen** [comment]
@havocp No problem about the package stuff! I'm going to rebase now. I just added a new commit adding a couple new tests and changing the way duplicate removal works in the ConfigDocument, and after lunch I'll get to work on addressing your comments.

**Comment by fpringvaldsen** [comment]
Thanks for the feedback as always!

**Comment by fpringvaldsen** [comment]
@havocp I've addressed your feedback!

**Comment by havocp** [comment]
Thanks!
